hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+  NotElem s '[] = ()
+
+type HasProperty s k t r = (k ~ PropertyKey s t, Elem s r, FindKey s r ~ t, KnownSymbol s)
+
+-- ---------------------------------------------------------------------
+
+emptyProperties :: Properties '[]
+emptyProperties = Properties Map.empty
+
+insert ::
+  forall s k r t.
+  (k ~ PropertyKey s t, NotElem s r, KnownSymbol s) =>
+  SPropertyKey k ->
+  MetaData t ->
+  Properties r ->
+  Properties (k ': r)
+insert key metadata (Properties old) = Properties (Map.insert (symbolVal (Proxy @s)) (SomePropertyKeyWithMetaData (key, metadata)) old)
+
+find ::
+  forall s k r t.
+  (HasProperty s k t r) =>
+  Properties r ->
+  (SPropertyKey k, MetaData t)
+find (Properties p) = case p Map.! symbolVal (Proxy @s) of
+  (SomePropertyKeyWithMetaData x) ->
+    -- Note [Constraints]
+    -- It's safe to use unsafeCoerce here:
+    --   Since each property name is unique that the redefinition will be prevented by predication on the type level list,
+    --   the value we get from the name-indexed map must be exactly the singleton and metadata corresponding to the type.
+    -- We drop this information at type level: some of the above type families return '() :: Constraint',
+    -- so GHC will consider them as redundant.
+    -- But we encode it using semantically identical 'Map' at term level,
+    -- which avoids inducting on the list by defining a new type class.
+    unsafeCoerce x
+
+-- ---------------------------------------------------------------------
+
+-- | Given the name of a defined property, generates a JSON parser of 'plcConfig'
+usePropertyEither ::
+  forall s k t r.
+  (HasProperty s k t r) =>
+  Properties r ->
+  A.Object ->
+  Either String (ToHsType t)
+usePropertyEither p = parseProperty @s (find p)
+
+-- | Like 'usePropertyEither' but returns 'defaultValue' on parse error
+useProperty ::
+  forall s k t r.
+  (HasProperty s k t r) =>
+  Properties r ->
+  Maybe A.Object ->
+  ToHsType t
+useProperty p = maybe (defaultValue metadata) (fromRight (defaultValue metadata) . usePropertyEither @s p)
+  where
+    (_, metadata) = find @s p
+
+parseProperty ::
+  forall s k t.
+  (k ~ PropertyKey s t, KnownSymbol s) =>
+  (SPropertyKey k, MetaData t) ->
+  A.Object ->
+  Either String (ToHsType t)
+parseProperty km x = case km of
+  (SNumber, _) -> parseEither
+  (SString, _) -> parseEither
+  (SBoolean, _) -> parseEither
+  (SObject, _) -> parseEither
+  (SArray, _) -> parseEither
+  (SEnum, EnumMetaData {..}) ->
+    A.parseEither
+      ( \o -> do
+          txt <- o A..: keyName
+          if txt `elem` enumValues
+            then pure txt
+            else fail $ "unknown enum option: " <> T.unpack txt
+      )
+      x
+  _ -> error "impossible"
+  where
+    keyName = T.pack $ symbolVal (Proxy @s)
+    parseEither :: forall a. A.FromJSON a => Either String a
+    parseEither = A.parseEither (A..: keyName) x
+
+-- ---------------------------------------------------------------------
+
+defineNumberProperty ::
```

Comment:
My understanding here is that you're supposed to type-apply `s`? Definitely needs haddock. 

Something to consider --- it might be nicer to give an overloaded label interface rather than force type-applications.

Reply:
Yeah, the explicit quantification is for type application of `s`